### ISSUE TITLE
[RFC][Serve][llm] ObjectRef-Based Weight Transfer for SGLang in Ray

### ISSUE THREAD

**Issue by SohamRajpure** [ray-project/ray#63120]
### Description

The goal is to surface the right questions. Most of the interesting decisions are in section 4. If you read only one section, read that one.

If something feels missing, it probably belongs in section 4 as another open question rather than as a resolved design choice.

I refer to _LMSYS_ and _HF survey_ throughout, jump to end for links.

## 1. Motivation
Every Ray-orchestrated async RL framework faces the same friction at the trainer-inference weight-sync boundary:

- Trainers produce updated weights as PyTorch tensors that already live somewhere in the cluster (usually owned by a Ray actor)

- SGLang’s existing weight-update APIs (`update_weights_from_[distributed, disk, tensor]`) all require the caller to bring its own transport: build a NCCL group, write a checkpoint, or be colocated.

- None of them accept a Ray `ObjectRef` directly. So users end up bridging Ray’s actor model to NCCL’s collective model by hand, and the two have no shared notion of object lifetime, placement, or failure. 

The LMSYS P2P work showed that RDMA via Mooncake gets weight-update times for Kimi-K2 from ~53s to ~7s. That work is transport-level. The gap this RFC is about is one layer up: a Ray-native API that lets a caller hand SGLang an `ObjectRef` and let the server figure out the rest.

## 2. What we want, roughly

```python
weight_ref = ray.put(state_dict_shard)
ray.get(sglang_server.update_weights_from_object_ref.remote(weight_ref, version=42))
```

That’s the whole user story. The hard parts are everything underneath.

## 3. Strawman
To give the discussion something concrete to push against, here is one possible v1. None of these choices are committed:

- **Refs are chunked:** caller passes `List[ObjectRef]`, one ref per all-gather bucket. Reject monolithic mega-refs in v1.

- **Wrapper holds the refs** in actor state until the corresponding bucket is loaded into the engine, then drops them. 

- **Pause window** is the same as LMSYS: `pause` → `load all buckets` → `post-process` → `version-bump` → `continue`

- **Version is monotone:** reject `v <= current`

- **No PD-dissagregation handling, no LoRA path, no fallback transport.** All deferred

This strawman is deliberately minimal. The point is to make the open questions in section 4 feel concrete.

## 4. Open Questions

These are decisions that need deliberation. Roughly ordered by how much disagreement I expect.

### 4.1 What's the relationship between this API and Ray Direct Transport?

This is probably the single biggest decision in the RFC, and it largely shapes the answers to the questions that follow.

Today, an `ObjectRef` to a GPU tensor implies a plasma roundtrip (serialize, copy through shared memory, deserialize on the other side). With RDT, the ref stays GPU-resident and dereferencing triggers a direct transfer.

It potentially subsumes section 4.2 (transport selection). If RDT handles NCCL/NIXL/Gloo dispatch internally, then "do we ship Mooncake-only or Mooncake-plus-fallbacks" stops being our question, Ray makes that choice. Our question becomes "do we let RDT pick, or do we override it for known-good paths like Mooncake-RDMA?"

Three positions:

- **RDT-first:** treat update_weights_from_object_ref as a thin wrapper over RDT-managed transfer. Mooncake becomes a backend that RDT may or may not select. We get NCCL/NIXL/Gloo for free as RDT matures, and the API ages well as Ray's transport story evolves.

- **Mooncake-first, RDT-later:** ship the LMSYS-validated path now, plan to migrate to RDT once it's proven at our scale. Defensible because Mooncake has demonstrated 1T params at 32 nodes; RDT's track record at that scale is unproven as of this writing.

- **Parallel paths:** expose both, let users pick. Most flexible, most maintenance burden.

The case for RDT-first is the strongest in principle, as we are making a Ray-native wrapper. SkyRL is also pushing RDT.

A version of "Mooncake-first, RDT-later" that I think is defensible: design the API surface so that switching the underlying transport from Mooncake to RDT is a non-breaking change. That means not exposing Mooncake-specific concepts (CPU replicas, Transfer Engine handles, P2P plans) in the public API, even as optional knobs.

This should be discussed before anything else in 4.

### 4.2 If we don’t go RDT-first, should v1 support more than one transport, or just Mooncake?

This question is condition on section 4.1. If RDT-first wins, much of it dissolves. Assuming Mooncake-first:

**For one transport:** simpler to ship, simpler to debug, LMSYS already validated Mooncake at scale. Adding NCCL fallback / CUDA IPC / ZMQ shm is the kind of generality that sounds free but will likely be time consuming.

**For multiple:** Mooncake requires RDMA NICs. Plenty of users (single-node setups, cloud envs without IB) won’t have them. Shipping with no fallback means those users keep using `update_weights_from_distributed` and never adopt the new API at all.

**Sub-question:** if we do support multiple, is selection per-update (one transport per call) or per-chunk (mix within a call)? Per-chunk is more flexible but the failure-mode space gets ugly fast.

### 4.3 How should ref lifetime be managed?

Ray’s `.remote()` future already keeps the ref alive while the call is in flight. Do we need additional holds inside the SGLang actor?

- **Yes, hold them:** defensive against weird ownership-transfer corners and against the caller cancelling early.

- **No, don’t:** Simpler, fewer cleanup paths, trust Ray’s refcounter to do its job.

I am leaning “yes” out of paranoia. Not sure if that is right.

### 4.4 Chunked vs. Monolithic refs – should monolithic be supported?

Chunked is clearly better for pipelining. But Atropos and ART hand around whole shards today. Forcing them to refactor before they can use the new API is a real adoption tax.

Options:
1. Reject monolithic in v1
2. Accept it but treat it as a single-chunk plan (leaning)
3. Accept it and warn loudly

### 4.5 What’s the right default chunk size?

LMSYS uses bucketed all-gather but doesn’t pin a number. Ray’s plasma store has soft limits around 100MB. NCCL likes ~1 GB buckets. Mooncake doesn’t care.

I have no data. This needs a benchmark, not a default from intuition. 

### 4.6 Concurrent-generation semantics during the pause

Three modes the surveyed libraries use:

- **Drain** (AReal, default LMSYS): finish in-flight requests, then pause

- **Abort** (PRIME-RL group cancellation, SkyRL prefix-resume): cancel in-flight, swap, retry

- **Inflight-continue** (PipelineRL): swap weights between forward passes, sequences see a mix.

Drain is the obvious v1 default. The question is whether we expose the others as configuration or wait until a user asks.

If we expose `inflight-continue`, there is a real semantic question: RDMA writes touch many parameters non-atomically, so a forward pass during the swap genuinely sees corrupted state. PipelineRL avoids this by holding a per-forward-pass lock; with RDMA we can’t, easily. That probably means inflight-continue is incompatible with the Mooncake transport, which is awkward if Mooncake is our only transport.

### 4.7 Where does PD disaggregation fit?
PD-aware sync has version-skew correctness implications (rollout buffer needs to tag both prefill and decode version for IS correction) that go well beyond this RFC. 

Options:

1. ignore PD entirely in v1, document fan-out is caller’s job

2. provide a fan-out helper but no version-skew support

3. full PD-aware design

I think (1) is right. Fan-out is one line of caller code. Version-skew is separate RFC.

### 4.8 What’s the failure contract?

Specifically: if the load fails partway through do we leave the engine paused or attempt recovery?

We could never resume on partial load, paused-and-silent-wrong. That’s defensible but it puts every failure on the caller’s shoulders. An alternative is: keep a checkpoint of the previous version’s weights and roll back on failure. That doubles GPU memory for the duration of the update, which may be unacceptable. 

This needs a real discussion with someone who has operated these systems at scale. I present no opinion here.

### 4.9 Is `version` enough metadata, or do we need more?

This is here just to avoid API signature churn (can always add Optionals)

A monotone integer is what every async RL library tracks. But the DeepSeek-V3.2 work on routing consistency (Keep Routing, Keep Sampling Mask) suggests that the inference-training metadata channel needs to be richer. Should `update_weights_from_object_ref` accept (or return) routing/sampling metadata or is that strictly a separate API?

Probably separate, but worth flagging here.

### 4.10 LoRA / adapter-only updates (v1, v2, or never?)

The HF survey calls LoRA-aware sync the “emerging differentiator” for MoE post-training. The transfer is so small (~50 MB) that it works on any transport, so the question isn’t technical complexity, it’s whether the parameter map needs an adapter-aware path and whether it is worth doing now or after we’ve seen the full-weight path land.

Need LoRA-using user input.

## 5. Things I’m not raising here

A few decisions I think aren’t worth discussing right now:

- Exact shape of WeightUpdateResult return value

- Cross-cluster transfer; probably never but here is a cool paper [https://arxiv.org/html/2604.15039]

If something feels missing, it probably belongs in section 4 as another open question rather than as a resolved design choice.

## Related Issues
- #62796 : SGLang + Ray Serve LLM Roadmap
- #62794 : RL weight sync and inference-trainer coordination
- vllm-project/vllm/issues/31848 : vLLM weight sync effort
- NovaSky-AI/SkyRL/issues/977 : SkyRL

## CC List
@eicherseiji @jeffreywang-anyscale @kouroshHakha 

### Link

- LMSYS P2P weight update [https://www.lmsys.org/blog/2026-04-29-p2p-update/]
- HF survey of 16 async RL libraries [https://huggingface.co/blog/async-rl-training-landscape]
